Add Weekly Talent Materials

lauslim12 opened the pull request:
## Problem

Issue number #36 (lack of weekly boss materials).

## What I Added

- All of the weekly talent materials, plus I have already run the `node genassets`
- Added drops to the trounce domains (except the Golden House, will do that in a separate PR)
- Assigned each character their weekly material (in the `talentMaterialWeekly` column, as requested)
- Added a small snippet to also display the weekly materials

## Results

With Diluc:

![image](https://user-images.githubusercontent.com/31909304/105194613-5aa09100-5b6c-11eb-9bae-f08e7ac2e45c.png)

With Keqing:

![image](https://user-images.githubusercontent.com/31909304/105194747-7e63d700-5b6c-11eb-97f5-94f7f006b700.png)

With Ayaka (should not be displayed because she does not have any weeklies):

![image](https://user-images.githubusercontent.com/31909304/105194802-8f144d00-5b6c-11eb-8bd5-fc943e962194.png)

With Ganyu (have not added the Golden House yet, will do in a separate PR after this is done):

![image](https://user-images.githubusercontent.com/31909304/105194918-9f2c2c80-5b6c-11eb-977d-64c4144949f0.png)

## Bugs

When clicking on the `Show on Schedule` button, the talent level-up material will also get clicked 😅. I suggest to pass a different prop or create a new checkbox whose purpose is to add the weekly boss to the list of tasks to be done.

EDIT: After reading the codebase further, I believe we can also make use of the global state variable which is the `Config` object at the `utils/configs.ts` file. We can add a new subtype inside the `Config` object, let's say `weeklyBosses`, and it will be an array which stores the strings `Dvalin`, `Andrius`, and `Childe`. The strings there will be used in the `Home/DomainView/index.tsx` file to render the stored 'bosses'.

Then, in order to trigger the storage, the easiest way is that we have to make a new component, let's say `ToggleWeekly.tsx`. This component will be exactly the same as `Toggle.tsx`, with the difference only being that the `useConfig` that we use will be the `weeklyBosses`. 

Well, actually we do not need to create a new component, just an `if` statement in the `Toggle.tsx` to differentiate it from the talent materials will suffice. In the `MaterialDisplay` component, we can pass a new prop, let's say `isWeekly` to the `Toggle` component. If `isWeekly` is `true`, then we will render the checkboxes for the bosses. If `isWeekly` is `false`, then we will render the checkboxes for `Common Ascension Materials`.

Still, in this case, making a new component is probably easier rather than passing props around.

## Closing

Thank you!

Resolves #36

lauslim12 committed:
Add weekly talent materials and run genassets

lauslim12 committed:
Add weekly talent materials in the DB

lauslim12 committed:
Add drops to trounce domains

lauslim12 committed:
Add weekly materials to characters

lauslim12 committed:
Display weekly talent material for each character

luaneko commented on web/db/characters.ts:
```diff
@@ -32,6 +49,7 @@ export type Character = {
   type: "Character";
   name: string;
   wiki: string;
+  talentMaterialWeekly: TalentMaterial[] | null; // Ayaka and Xiao does not have weeklies yet.
```
Is there a particular reason you went with `TalentMaterial[] | null`? I think an empty array would be good enough to indicate the lack of weekly material.

lauslim12 replied:
My bad, didn't cross my mind yesterday. Sorry! Fixed!

luaneko commented on web/pages/characters/[name].tsx:
```diff
@@ -57,6 +57,12 @@ const CharacterInfo = ({ data, name }: Props) => {
 
+            {/* If null, don't do anything. Refer to 'db/characters.tsx' in the Weekly Materials. */}
+            {/* The 'null' value is possible on unreleased characters. We have no idea about their weeklies. */}
+            {character.talentMaterialWeekly?.map((material) => (
```
Same comment as above. `map` would return an empty array which doesn't render anything anyway so I don't see the benefit of null.

lauslim12 replied:
This too! Sorry! Resolved.

luaneko reviewed (changes requested):
This is mostly good except for one small typing issue.

luaneko commented:
> I believe we can also make use of the global state variable which is the Config object at the utils/configs.ts file.

To be precise it's not a global state variable, it's passed on via a [React context](https://github.com/chiyadev/genshin-schedule/blob/master/web/components/ConfigsProvider.tsx#L19) :) I don't think there's any global state in this codebase.

> We can add a new subtype inside the Config object, let's say weeklyBosses, and it will be an array which stores the strings Dvalin, Andrius, and Childe. The strings there will be used in the Home/DomainView/index.tsx file to render the stored 'bosses'.

Instead of storing boss names, we should store the names of characters which require such bosses, so the field would be named like `charactersWeekly`.

> In the MaterialDisplay component, we can pass a new prop, let's say isWeekly to the Toggle component. If isWeekly is true, then we will render the checkboxes for the bosses. If isWeekly is false, then we will render the checkboxes for Common Ascension Materials.

We can simplify that by simply changing this line

```jsx
const [list, setList] = useConfig("characters");
```

like this

```jsx
const [list, setList] = useConfig(isWeekly ? "charactersWeekly" : "characters");
```

Then there would be no need for a new component and duplicated code. The `isWeekly` prop would be passed to `MaterialDisplay` which just passes it onto `Toggle`.

lauslim12 committed:
Change null into empty array

lauslim12 committed:
Add new 'charactersWeekly' context

lauslim12 committed:
Add new prop to differentiate talent materials checked

lauslim12 committed:
Modified 'useConfig' to also use the new context

lauslim12 commented:
> To be precise it's not a global state variable, it's passed on via a React context :) I don't think there's any global state in this codebase.

My bad! Sorry!

As for your requested changes, I have done them as requested, and I have also made it so that when we click on one checkbox, the other does not get clicked too (the bug has been resolved).

![image](https://user-images.githubusercontent.com/31909304/105274734-61f98600-5bd0-11eb-989d-562a84b318a5.png)

I have a question before rendering them on the main page though. What do you think is the best way to display the weekly domains?

Should we do it like this?

![image](https://user-images.githubusercontent.com/31909304/105275122-2b703b00-5bd1-11eb-9a76-73fdbb5c06c8.png)

~~If we do it like this, there appears a bug again where the weekly domain does not appear if the current day is not the talent day of the character. Let me try something along the way.~~

EDIT 1: Nevermind, I fixed the bug already and it has rendered properly.

![image](https://user-images.githubusercontent.com/31909304/105275996-067cc780-5bd3-11eb-80ed-ebfe47e51214.png)

EDIT 2: We know that the character's card will be darkened a bit when we add the normal talent material to the list of tasks to be done. The question is, should I also add the darkened effect if we have added the weekly talent material to the list of tasks to be done? Right now, when we add the weekly talent material to the list of tasks, the card does not get darkened, it stays the same color, as seen in below picture.

![image](https://user-images.githubusercontent.com/31909304/105276234-78eda780-5bd3-11eb-9285-79efc9c7a1fc.png)

Thank you!

luaneko commented:
> should I also add the darkened effect if we have added the weekly talent material to the list of tasks to be done?

Yup, let's make it dark when it's scheduled for either talent or weekly materials.

lauslim12 committed:
Render weekly bosses propely in main page

lauslim12 committed:
Add darken effect for either weekly or normal talent chosen

lauslim12 commented:
All right! All done!

![image](https://user-images.githubusercontent.com/31909304/105278655-8c4f4180-5bd8-11eb-8be9-bfccf8a68c32.png)

![image](https://user-images.githubusercontent.com/31909304/105278703-9f621180-5bd8-11eb-9262-6fea9f9a324d.png)

luaneko commented on web/components/Customize/CharacterList/Icon.tsx:
```diff
@@ -7,7 +7,10 @@ import NextLink from "next/link";
-  const [existing] = useConfig("characters");
+  const [charactersWeekly] = useConfig("charactersWeekly");
+  const [characters] = useConfig("characters");
+  const existing = [...new Set([...charactersWeekly, ...characters])];
```
This line instantiates a new array every render, making the `useMemo` below pointless. The set should be moved inside the memo and call `has` on it instead of converting it to array in a roundabout way.

luaneko commented on web/components/Home/DomainView/index.tsx:
```diff
@@ -44,6 +44,7 @@ const DomainView = () => {
   const [characters] = useConfig("characters");
   const [weapons] = useConfig("weapons");
   const [artifacts] = useConfig("artifacts");
+  const [weeklyBosses] = useConfig("charactersWeekly");
```
I think you forgot to rename this to `charactersWeekly`.

luaneko reviewed (changes requested):
Almost there!

lauslim12 committed:
Fix typo on the 'useConfig'

lauslim12 committed:
Fix 'useMemo' minor error

lauslim12 commented:
Okay! Done as requested!

Commit:
Merge remote-tracking branch 'upstream/master' into weekly-talent-materials

# Conflicts:
#	web/components/Home/DomainView/index.tsx
#	web/utils/configs.ts

Commit:
Add changelog entry for weekly talent materials addition

luaneko commented:
Thanks a lot!
